specs/core/1_beacon-chain-misc.md

```diff
@@ -0,0 +1,234 @@
+            - [`BeaconState`](#beaconstate)
+            - [`BeaconBlockBody`](#beaconblockbody)
+        - [Persistent committees](#persistent-committees)
+        - [Shard receipt processing](#shard-receipt-processing)
+
+<!-- /TOC -->
+
+## Configuration
+
+| Name | Value | Unit | Duration
+| - | - | - | - |
+| `MAX_SHARD_RECEIPT_PROOFS` | `2**0` (= 1) | - | - |
+| `PERIOD_COMMITTEE_ROOT_LENGTH` | `2**8` (= 256) | periods | ~9 months |
+| `MINOR_REWARD_QUOTIENT` | `2**8` (=256) | - | - |
+
+## Containers
+
+#### `CompactCommittee`
+
+```python
+class CompactCommittee(Container):
+    pubkeys: List[BLSPubkey, MAX_VALIDATORS_PER_COMMITTEE]
+    compact_validators: List[uint64, MAX_VALIDATORS_PER_COMMITTEE]
+```
+
+#### `ShardReceiptProof`
+
+```python
+class ShardReceiptProof(Container):
+    shard: Shard
+    proof: List[Hash, PLACEHOLDER]
+    receipt: List[ShardReceiptDelta, PLACEHOLDER]
+```
+
+## Helper functions
+
+#### `pack_compact_validator`
+
+```python
+def pack_compact_validator(index: int, slashed: bool, balance_in_increments: int) -> int:
+    """
+    Creates a compact validator object representing index, slashed status, and compressed balance.
+    Takes as input balance-in-increments (// EFFECTIVE_BALANCE_INCREMENT) to preserve symmetry with
+    the unpacking function.
+    """
+    return (index << 16) + (slashed << 15) + balance_in_increments
+```
+
+#### `unpack_compact_validator`
+
+```python
+def unpack_compact_validator(compact_validator: int) -> Tuple[int, bool, int]:
+    """
+    Returns validator index, slashed, balance // EFFECTIVE_BALANCE_INCREMENT
+    """
+    return compact_validator >> 16, bool((compact_validator >> 15) % 2), compact_validator & (2**15 - 1)
+```
+
+#### `committee_to_compact_committee`
+
+```python
+def committee_to_compact_committee(state: BeaconState, committee: Sequence[ValidatorIndex]) -> CompactCommittee:
+    """
+    Given a state and a list of validator indices, outputs the CompactCommittee representing them.
+    """
+    validators = [state.validators[i] for i in committee]
+    compact_validators = [
+        pack_compact_validator(i, v.slashed, v.effective_balance // EFFECTIVE_BALANCE_INCREMENT)
+        for i, v in zip(committee, validators)
+    ]
+    pubkeys = [v.pubkey for v in validators]
+    return CompactCommittee(pubkeys=pubkeys, compact_validators=compact_validators)
+```
+
+#### `verify_merkle_proof`
+
+```python
+def verify_merkle_proof(leaf: Hash, proof: Sequence[Hash], index: GeneralizedIndex, root: Hash) -> bool:
+    assert len(proof) == get_generalized_index_length(index)
+    for i, h in enumerate(proof):
+        if get_generalized_index_bit(index, i):
+            leaf = hash(h + leaf)
+        else:
+            leaf = hash(leaf + h)
+    return leaf == root
+```
+
+#### `compute_historical_state_generalized_index`
+
+```python
+def compute_historical_state_generalized_index(earlier: ShardSlot, later: ShardSlot) -> GeneralizedIndex:
```

Comment:
I put up a PR with an addition to `history_acc` and how it is computed to make this work - https://github.com/ethereum/eth2.0-specs/pull/1379

Reply:
I simulated the algorithm with the stated change, using the following code:
```python3
def compute_historic_accumulator(slot):
    slots = [None]*14
    for i in range(1, slot + 1):
        compute_historic_accumulator_per_slot(i, slots)
    print(slots)



def compute_historic_accumulator_per_slot(slot, slots):
    depth = 63
    while depth >= 0:
        if slot % 2**depth == 0:
            slots[depth] = slot
            break
        depth -= 1


def compute_historical_state_generalized_index(earlier, later):
    for i in range(63, -1, -1):
        if (later - 1) & 2**i > (earlier - 1) & 2**i:
            later = later - ((later - 1) % 2**i) - 1
            print(i)
            print(later)
            compute_historic_accumulator(later)




compute_historic_accumulator(2393943)
compute_historical_state_generalized_index(2342, 2393943);
```
Results of the simulation:
```python3
steps to solution = 17
earlier = 2342
later = 2393943

[2393943, 2393942, 2393940, 2393928, 2393936, 2393888, 2393920, 2393728, 2393856, 2393600, 2393088, 2390016, 2387968, 2383872, 2375680, 2392064, 2293760, 2228224, 2359296, 1572864, 1048576, 2097152]
i = 21
later = 2097152
[2097151, 2097150, 2097148, 2097144, 2097136, 2097120, 2097088, 2097024, 2096896, 2096640, 2096128, 2095104, 2093056, 2088960, 2080768, 2064384, 2031616, 1966080, 1835008, 1572864, 1048576, 2097152]
i = 20
later = 1048576
[1048575, 1048574, 1048572, 1048568, 1048560, 1048544, 1048512, 1048448, 1048320, 1048064, 1047552, 1046528, 1044480, 1040384, 1032192, 1015808, 983040, 917504, 786432, 524288, 1048576]
i = 19
later = 524288
[524287, 524286, 524284, 524280, 524272, 524256, 524224, 524160, 524032, 523776, 523264, 522240, 520192, 516096, 507904, 491520, 458752, 393216, 262144, 524288]
i = 18
later = 262144
[262143, 262142, 262140, 262136, 262128, 262112, 262080, 262016, 261888, 261632, 261120, 260096, 258048, 253952, 245760, 229376, 196608, 131072, 262144]
i = 17
later = 131072
[131071, 131070, 131068, 131064, 131056, 131040, 131008, 130944, 130816, 130560, 130048, 129024, 126976, 122880, 114688, 98304, 65536, 131072]
i = 16
later = 65536
[65535, 65534, 65532, 65528, 65520, 65504, 65472, 65408, 65280, 65024, 64512, 63488, 61440, 57344, 49152, 32768, 65536]
i = 15
later = 32768
[32767, 32766, 32764, 32760, 32752, 32736, 32704, 32640, 32512, 32256, 31744, 30720, 28672, 24576, 16384, 32768]
i = 14
later = 16384
[16383, 16382, 16380, 16376, 16368, 16352, 16320, 16256, 16128, 15872, 15360, 14336, 12288, 8192, 16384]
i = 13
later = 8192
[8191, 8190, 8188, 8184, 8176, 8160, 8128, 8064, 7936, 7680, 7168, 6144, 4096, 8192]
i = 12
later = 4096
[4095, 4094, 4092, 4088, 4080, 4064, 4032, 3968, 3840, 3584, 3072, 2048, 4096]
i = 10
later = 3072
[3071, 3070, 3068, 3064, 3056, 3040, 3008, 2944, 2816, 2560, 3072, 2048]
i = 9
later = 2560
[2559, 2558, 2556, 2552, 2544, 2528, 2496, 2432, 2304, 2560, 1024, 2048]
i = 7
later = 2432
[2431, 2430, 2428, 2424, 2416, 2400, 2368, 2432, 2304, 1536, 1024, 2048]
i = 6
later = 2368
[2367, 2366, 2364, 2360, 2352, 2336, 2368, 2176, 2304, 1536, 1024, 2048]
i = 4
later = 2352
[2351, 2350, 2348, 2344, 2352, 2336, 2240, 2176, 2304, 1536, 1024, 2048]
i = 3
later = 2344
[2343, 2342, 2340, 2344, 2320, 2336, 2240, 2176, 2304, 1536, 1024, 2048]
i = 1
later = 2342
[2341, 2342, 2340, 2328, 2320, 2336, 2240, 2176, 2304, 1536, 1024, 2048]
```

TLDR - it looks like the solution can be slightly optimized since the first step does not pick the optimal index. In this case, it chose `i = 21` which equated to `2097152` when it could have just chosen `i = 20` which equated to `1048576` in the first step.

Also curious - what the name of this exact algorithm in this case?

Reply:
It's basically putting the most recent state at slot `k * 2**i` in position `i`.